Add support for Debian 9 aka. stretch

SleuthKid opened the pull request:
Adding support for the Debian 9 (stretch) build environment.

REF: apache/couchdb#781

rajudev commented on ansible/debian-9-erlang-18.3.yml:
```diff
@@ -0,0 +1,24 @@
+
+- hosts: couchdb-ci-worker
+  vars:
+    erlang_solutions_pkg_ver: "1:18.3-1"
```
Why not do it for the for the version of erlang in stable. which is 19.2 
instead of 18.3 

https://tracker.debian.org/pkg/erlang

SleuthKid replied:
As far as I understand there are two different builds. One with 18.3 (LTS?!) and one with the distribution default erlang (ansible/debian-9-erlang-default.yml) , 19.2 in that case.

rajudev replied:
As far as I understand, the number 18.3 refers to the version of erlang in that particular distro. which in case of stretch is 19.2 . I may be wrong. but that is what I infer from the situation.

wohali replied:
We currently build against 18.3 in our CI pipeline and for released packages. When we move up to 19, we'll move to 19 across all releases, not just this one.

rajudev replied:
understood. Thanks for the explanation.

rajudev commented:
Thanks a lot  @SleuthKid for the Pull request, Me and friends were about to start working #781  , and here you sent the PR already. But it seems you'r builds are failing. Lets debug that. Hope we see the deb for stretch at the earliest. More power to you.

SleuthKid commented:
@rajudev looking at the commit log the travis-ci pipeline was never functional for this repository. I think @wohali is still working on it.

wohali commented:
The Travis CI integration needs to be disabled for this repo. I'll file a ticket with Apache Infra to do so.

rajudev commented:
@wohali Why do you need to disable Travis CI Integration for this repo?

wohali commented:
@SleuthKid Can you please change the comment in your commit to include another line with the phrase: `Closes #3`  Just edit the commit and force-push to your branch, please.

wohali commented:
@rajudev I edited the comment already - Travis can test the repo but our official CI pipeline for building packages is Apache-hosted Jenkins CI, for policy and security reasons.

rajudev commented:
Thanks @wohali for the explanation. I see, https://builds.apache.org/job/CouchDB/ 
How often are the builds run on the Apache build server?
How long will it take for the built deb to be pushed to https://apache.bintray.com/couchdb-deb/ ?

wohali commented:
Jenkins CI runs on all checkins to CouchDB for the `master`, `#.#.#`, and any `jenkins-*` branches.

The steps to releasing binaries for a new platform are as follows:

1. I test, approve and merge this PR
2. I update the `Jenkinsfile` in `apache/couchdb` to include the new Debian 9 build, first testing on a special jenkins-debian-9 branch, then on master as well
3. I confirm that Jenkins is successfully building the new platform's packages correctly
4. I test the packages that Jenkins builds automatically for validity on a Debian 9 machine - following our release acceptance policy procedure
5. I upload those packages to bintray.

With luck it'll be done by the end of this weekend.

wohali commented on bin/debian-9-base/create-container.sh:
```diff
@@ -0,0 +1,28 @@
+#!/usr/bin/env bash
```
None of these .sh scripts are set `chmod +x`. Please do so.

SleuthKid replied:
Good catch. Solved with 7fb7bc9

SleuthKid replied:
Debian provides the libmozjs185-1.0 package with the same spidermonkey version. Could we just skip the source build and use that?

wohali replied:
No. We've found some serious segfaults in the libmozjs185-1.0 package and build our own with different flags to avoid these.

SleuthKid replied:
This bug is actually fixed in the debian package: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=811665

Besides GCC6 prints a lot of possible errors (warnings) in the source code that are partly addressed by the debian packaging team.

I would recommend using the official package - at least for the stretch build as the root of your problems with seem to come from building for arm?

https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=839048

SleuthKid commented:
I found another possible show stopper:

The erlang repository for stretch seems to not provide 18.3 packages in the way the playbook wants them:

http://packages.erlang-solutions.com/debian/dists/stretch/contrib/binary-amd64/Packages

rajudev commented:
@SleuthKid You need to pull erlang 18.3 from the deb repos of erlang instead that of debian's. @wohali mentioned this yesterday in the IRC chat, yesterday. 
let me find the repo url in the mean time

wohali commented:
@SleuthKid Yup, you're going to have to just install `esl-erlang:18.3.4.5` it looks like.

rajudev commented:
@SleuthKid The repos are here https://packages.erlang-solutions.com/erlang/

SleuthKid commented:
I fixed the spidermonkey build with the latest commit by porting the debian patch to the playbook role.

I fixed the 18.3 erlang container build by using the esl-erlang:18.3.4.5 package as suggested.

wohali commented on ansible/roles/spidermonkey/tasks/main.yml:
```diff
@@ -19,19 +19,30 @@
+    src: "{{ role_path }}/files/fix-811665.patch"
+    basedir: /opt/js-1.8.5/
+    strip: 1
+  when: ansible_distribution == "Debian"
```
This needs to be conditional to only apply this patch for the Debian 9 build, not all builds please.

SleuthKid replied:
Done :+1:

SleuthKid commented:
@wohali are we ready to merge?

wohali commented:
Sorry for the delays. I need to run the test build by hand still, and life has been busy the past week.

I will try and slot time in for this tonight.

wohali commented:
Progress, note a slight modification:

- [X] I test, approve and merge this PR
- [X] I update the Jenkinsfile in apache/couchdb to include the new Debian 9 build, first testing on a special jenkins-debian-9 branch.
- [ ] I confirm that Jenkins is successfully building the new platform's packages correctly
- [ ] I test the packages that Jenkins builds automatically for validity on a Debian 9 machine, following our release acceptance policy procedure
- [ ] Merge the Jenkinsfile to master
- [ ] Once 2.1.1 is released (imminent) I upload Debian 9 packages to bintray along with the other releases.
